### PR TITLE
Bump smdebug version to 1.0.29

### PR DESCRIPTION
**Pull request by malav-shastri**
*Issue #, if available:*
*Description of changes:*
- Debugger team released a fix for supporting XgBoost report and other debugger functionality for XgBoost version 1.3 and beyond - https://github.com/awslabs/sagemaker-debugger/pull/648
- Bumping the version of smdebug to 1.0.29 to patch the fix in XgBoost container 1.3-1
- Made related changes to support the new smdebug version
- Related PR -  https://github.com/awslabs/sagemaker-debugger/pull/650

*Testing:*
- Unit test
- Flake8
- Integration tests
- Tested notebook with the updated image https://github.com/aws/amazon-sagemaker-examples/blob/main/sagemaker-debugger/xgboost_training_report/higgs_boson_detection.ipynb

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docker/1.3-1/final/Dockerfile.cpu [hunk 1]
```diff
@@ -11,6 +11,10 @@ ARG SAGEMAKER_XGBOOST_VERSION
 COPY requirements.txt /requirements.txt
 RUN python3 -m pip install -r /requirements.txt && rm /requirements.txt
 
+# Install smdebug from source
+RUN python3 -m pip install git+https://github.com/awslabs/sagemaker-debugger.git@1.0.29
```
Comment (dewan-c) [inline docker/1.3-1/final/Dockerfile.cpu:15]: Curious, why are we installing from branch here if it's available through pip?

File: src/sagemaker_xgboost_container/callback.py [hunk 2]
```diff
@@ -37,9 +17,13 @@ def add_debugging(callbacks, hyperparameters, train_dmatrix,
...
+        hook = Hook.hook_from_config(json_config_path)
+        hook.hyperparameters = hyperparameters
+        hook.train_data = train_dmatrix
+        if val_dmatrix is not None:
+            hook.validation_data = val_dmatrix
+        callbacks.append(hook)
         logging.info("Debug hook created from config")
     except Exception as e:
         logging.debug("Failed to create debug hook", e)
-    else:
```
Comment (dewan-c) [inline src/sagemaker_xgboost_container/callback.py:44]: Nit - this else is now redundant.

File: docker/1.3-1/final/Dockerfile.cpu (hunk 1 above)
Comment (malav-shastri) [reply docker/1.3-1/final/Dockerfile.cpu:15]: so this new version is not available through pip, according to them they haven't released wheels recently and their pipeline for that is blocked. They are not sure when they would be able to resolve that and suggested me to directly install it from source as almost all of their dependencies are installing from source only. To unblock this issue for time being if there are no major cons to this approach maybe we can treat it as a temporary work around and can request them to release a new wheel eventually.

File: src/sagemaker_xgboost_container/callback.py (hunk 2 above)
Comment (malav-shastri) [reply src/sagemaker_xgboost_container/callback.py:44]: umm so currently I am not using it right?

File: docker/1.3-1/final/Dockerfile.cpu (hunk 1 above)
Comment (dewan-c) [reply docker/1.3-1/final/Dockerfile.cpu:15]: Ok, makes sense, thanks for the explanation.

File: src/sagemaker_xgboost_container/callback.py (hunk 2 above)
Comment (dewan-c) [reply src/sagemaker_xgboost_container/callback.py:44]: Why keep unused code? I meant removing this else altogether would work just the same.

**Review by dewan-c — APPROVED** [review]

File: src/sagemaker_xgboost_container/callback.py (hunk 2 above)
Comment (dewan-c) [reply src/sagemaker_xgboost_container/callback.py:44]: Ak it's my bad. I didn't notice that return was moved under try. Fyi- you don't need the `return` either.

**Pull request merged**